tests/beta/interfaces/test_tape_autograd.py

```diff
@@ -732,3 +730,558 @@ def cost(x, device):
+        """Test evaluation and Jacobian if there are no trainable parameters"""
+        dev = qml.device(dev_name, wires=2)
+
+        @qnode(dev, diff_method=diff_method, interface="autograd")
+        def circuit(a, b):
+            qml.RY(a, wires=0)
+            qml.RX(b, wires=0)
+            qml.CNOT(wires=[0, 1])
+            return expval(qml.PauliZ(0)), expval(qml.PauliZ(1))
+
+        a = np.array(0.1, requires_grad=False)
+        b = np.array(0.2, requires_grad=False)
+
+        res = circuit(a, b)
+
+        if diff_method == "finite-diff":
+            assert circuit.qtape.trainable_params == set()
+
+        assert res.shape == (2,)
+        assert isinstance(res, np.ndarray)
+
+        assert not qml.jacobian(circuit)(a, b)
+
+        def cost(a, b):
+            return np.sum(circuit(a, b))
+
+        with pytest.warns(UserWarning, match="Output seems independent of input"):
+            grad = qml.grad(cost)(a, b)
+
+        assert grad == tuple()
+
+    def test_matrix_parameter(self, dev_name, diff_method, tol):
+        """Test that the TF interface works correctly
+        with a matrix parameter"""
+        U = np.array([[0, 1], [1, 0]], requires_grad=False)
+        a = np.array(0.1, requires_grad=True)
+
+        dev = qml.device(dev_name, wires=2)
+
+        @qnode(dev, diff_method=diff_method, interface="autograd")
+        def circuit(U, a):
+            qml.QubitUnitary(U, wires=0)
+            qml.RY(a, wires=0)
+            return expval(qml.PauliZ(0))
+
+        res = circuit(U, a)
+
+        if diff_method == "finite-diff":
+            assert circuit.qtape.trainable_params == {1}
+
+        res = qml.grad(circuit)(U, a)
+        assert np.allclose(res, np.sin(a), atol=tol, rtol=0)
+
+    def test_differentiable_expand(self, dev_name, diff_method, mocker, tol):
+        """Test that operation and nested tapes expansion
+        is differentiable"""
+        mock = mocker.patch.object(qml.operation.Operation, "do_check_domain", False)
+
+        class U3(qml.U3):
+            def expand(self):
+                theta, phi, lam = self.data
+                wires = self.wires
+
+                with QuantumTape() as tape:
+                    qml.Rot(lam, theta, -lam, wires=wires)
+                    qml.PhaseShift(phi + lam, wires=wires)
+
+                return tape
+
+        dev = qml.device(dev_name, wires=1)
+        a = np.array(0.1, requires_grad=False)
+        p = np.array([0.1, 0.2, 0.3], requires_grad=True)
+
+        @qnode(dev, diff_method=diff_method, interface="autograd")
+        def circuit(a, p):
+            qml.RX(a, wires=0)
+            U3(p[0], p[1], p[2], wires=0)
+            return expval(qml.PauliX(0))
+
+        res = circuit(a, p)
+
+        if diff_method == "finite-diff":
+            assert circuit.qtape.trainable_params == {1, 2, 3, 4}
+        elif diff_method == "backprop":
+            assert circuit.qtape.trainable_params == {0, 1, 2, 3, 4}

```

Comment:
Yes, this one is quite subtle.

* The `QuantumTape.trainable_params` property is _only ever used_ to work out which parameters to loop over in the `QuantumTape.jacobian()` method.

* For a 'blackbox' device, we wrap the tape in an interface, register the custom gradient, and use tools provided by the interface (`should_record_backprop`, `requires_grad`) to work out which parameters are trainable, and which we need to differentiate within `QuantumTape.jacobian()`.

* For a 'whitebox' device, no interface wrapping is performed, and `QuantumTape.jacobian()` is never called. As a result, `QuantumTape.trainable_params` is never set --- the ML framework uses its own backprop logic and its own bookkeeping re: trainable parameters.

I'll add a comment to the test explaining this.

Reply:
Maybe to avoid confusion, we should set `trainable_params` to raise an exception for whitebox devices?

Reply:
Could it be worth setting `trainable_params` to `None` or to some other distinct value for 'whitebox' devices? If I understand correctly it would return incorrect output and could lead to conversion.

Reply:
I'm a bit hesitant to add additional logic to the `trainable_params` in the case of a backprop device, since `trainable_params` is not used for backprop devices. I've added a comment explaining this to the test.